### PR TITLE
[ddp] use named_params and named_buffers explicitly

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65181

This PR changes `state_dict()` during sync to `named_parameters` and `named_buffers` explicitly. the underlying motivation is that, `state_dict()` doesn't necessarily equals to "params + buffers" for all cases, state_dict is used for checkpoint purpose mainly, and params/buffers are used for training, we might have cases that params/buffers be in different forms with state_dict (i.e. state_dict we might want to save in small pieces of tensors while in training we want to concat the tensors together for performance reasons).

Differential Revision: [D31007085](https://our.internmc.facebook.com/intern/diff/D31007085/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D31007085/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

### CODE REVIEW

**Review by rohan-varma — COMMENTED** [review]
Looks good overall, but a couple of test failures that appear to be related.

**Comment by wanchaol** [comment]
> Looks good overall, but a couple of test failures that appear to be related.

Thanks yeah just fixed them, not detaching gradients in the last try, should be ready.

**Review by rohan-varma — APPROVED** [review]
Thanks!
